### PR TITLE
feat(vim-patch.sh): better detection of remote name

### PR DESCRIPTION
**Pull request by zeertzjq**
`vim-patch.sh` hard-codes `origin` as the remote name of fork, but some may prefer using `origin` for Neovim upstream and some other remote name for fork.

### CODE REVIEW

File: scripts/vim-patch.sh [hunk 1]
```diff
@@ -376,8 +376,13 @@ submit_pr() {
   pr_title="$(printf 'vim-patch:%s' "${pr_title#,}")"
 
   if [[ $push_first -ne 0 ]]; then
-    echo "Pushing to 'origin/${checked_out_branch}'."
-    output="$(git push origin "${checked_out_branch}" 2>&1)" &&
+    local checked_out_branch_remote
+    checked_out_branch_remote="$(git config --get "branch.${checked_out_branch}.remote")"
```
Comment (jamessan) [inline scripts/vim-patch.sh:380]: I guess I'm spoiled by [git hub](https://github.com/sociomantic-tsunami/git-hub/)'s behavior.

The way all of my clones are setup, there are two remotes -- `fork` and `upstream`.  A plain `git pull` will pull from the tracking branch in upstream and a plain `git push` will push to `fork`.

I usually create a branch with `git checkout -b foo upstream/master`.  Now, `git config --get branch.foo.remote` is going to show `upstream` for me, which isn't correct in your scenario.

Granted, when `git hub` is being used, `$push_first` is false.  However, to handle similar setups, `git config --get remote.pushdefault` should probably be checked first.

Comment (jamessan) [reply scripts/vim-patch.sh:380]: Actually, looking at `git-config(1)`, it looks like the precedence is:

1. `branch.<name>.pushRemote`
2. `remote.pushDefault`
3. `branch.<name>.remote`

Comment (zeertzjq) [reply scripts/vim-patch.sh:380]:
```
remote.pushDefault
The remote to push to by default. Overrides branch.<name>.remote for all branches, and is overridden by branch.<name>.pushRemote for specific branches.
``

`remote.pushDefault` has the lowest precedence.

Comment (zeertzjq) [reply scripts/vim-patch.sh:380]: Oh, I didn't know these two options before. Git sometimes also automatically sets `branch.<name>.remote`, so I guess I'll use only `branch.<name>.pushRemote` and `remote.pushDefault` instead.

File: scripts/vim-patch.sh [hunk 2]
```diff
@@ -376,8 +376,16 @@ submit_pr() {
...
 
   if [[ $push_first -ne 0 ]]; then
-    echo "Pushing to 'origin/${checked_out_branch}'."
-    output="$(git push origin "${checked_out_branch}" 2>&1)" &&
+    local push_remote
+    push_remote="$(git config --get branch."${checked_out_branch}".pushRemote)"
+    if [[ -z "$push_remote" ]]; then
+      push_remote="$(git config --get remote.pushDefault)"
+      if [[ -z "$push_remote" ]]; then
+        push_remote="origin"
```
Comment (jamessan) [inline scripts/vim-patch.sh:384]: Probably still makes sense to use `git config --get branch.<name>.remote` here.

---
Alternatively, would it be better to do something like `find_get_remote` but for the user's repo?

```sh
find_push_remote() {
  git_remote=$(git remote -v \
    | awk '$2 !~ /github.com[:\/]neovim\/neovim/ && $3 == "(push)" {print $1; exit}')
  if [[ -z "$git_remote" ]]; then
    git_remote="origin"
  fi
  echo "$git_remote"
}
```

Comment (zeertzjq) [reply scripts/vim-patch.sh:384]: Maybe `branch.<name>.remote` can used if it is not Neovim upstream remote.

File: scripts/vim-patch.sh [hunk 3]
```diff
@@ -124,13 +124,22 @@ commit_message() {
...
-  if [[ -z "$git_remote" ]]; then
-    git_remote="origin"
+  if [[ -z "$nvim_remote" ]]; then
+    nvim_remote="origin"
   fi
-  echo "$git_remote"
+  echo "$nvim_remote"
+}
+
+find_fork_remote() {
```
Comment (justinmk) [inline scripts/vim-patch.sh:136]: these functions are almost identical, can be easily parameterized. something like

```
local r
if [ "$1" = fork ] ; then
  r=$(git remote -v | awk '$2 !~ /github.com[:\/]neovim\/neovim/ && $3 == "(fetch)" {print $1; exit}')
else
  r=$(git remote -v | awk '$2 ~ /github.com[:\/]neovim\/neovim/ && $3 == "(fetch)" {print $1; exit}')
fi
```

**Review by justinmk — APPROVED** [review]
LGTM but I don't use the "push" feature of vim-patch.sh, waiting for others to approve/merge. cc @janlazo

**Review by jamessan — APPROVED** [review]
LGTM, but I can't test it currently

File: scripts/vim-patch.sh [hunk 4]
```diff
@@ -125,8 +125,12 @@ commit_message() {
 }
 
 find_git_remote() {
-  git_remote=$(git remote -v \
-    | awk '$2 ~ /github.com[:\/]neovim\/neovim/ && $3 == "(fetch)" {print $1; exit}')
+  local git_remote
+  if [[ "$1" == fork ]]; then
```
Comment (janlazo) [inline scripts/vim-patch.sh:129]: ./scripts/vim-patch.sh: line 129: $1: unbound variable

Comment (zeertzjq) [reply scripts/vim-patch.sh]: I changed it to `${1-}`.

**Comment by janlazo** [comment]
`vim-patch.sh -s` silently fails with exit code 1.

**Comment by zeertzjq** [comment]
> `vim-patch.sh -s` silently fails with exit code 1.

Under what circumstances?

**Comment by zeertzjq** [comment]
> `vim-patch.sh -s` silently fails with exit code 1.

I added `|| true`. Should be fine now.

**Pull request merged**